File: src/main/java/pw/mihou/nexus/features/command/interceptors/core/NexusCommandInterceptorCore.java
package pw.mihou.nexus.features.command.interceptors.core;

import pw.mihou.nexus.features.command.core.NexusMiddlewareEventCore;
import pw.mihou.nexus.features.command.facade.NexusCommandEvent;
import pw.mihou.nexus.features.command.interceptors.facades.*;
import pw.mihou.nexus.features.command.interceptors.repositories.NexusMiddlewareGateRepository;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NexusCommandInterceptorCore {

    private static final Map<String, NexusCommandInterceptor> interceptors = new HashMap<>();

    /**
     * Adds the middleware into the command interceptor storage.
     *
     * @param name The name of the middleware to add.
     * @param middleware The middleware functionality.
     */
    public static void addMiddleware(String name, NexusMiddleware middleware) {
        interceptors.put(name, middleware);
    }

    /**
     * Adds a repository of command interceptors to the command interceptor
     * storage.
     *
     * @param repository    The repository to add.
     */
    public static void addRepository(NexusInterceptorRepository repository) {
        repository.define();
    }

    /**
     * Adds the afterware into the command interceptor storage.
     *
     * @param name The name of the afterware to add.
     * @param afterware The afterware functionality.
     */
    public static void addAfterware(String name, NexusAfterware afterware) {
        interceptors.put(name, afterware);
    }

    /**
     * An internal method that is used to execute a command interceptor based on its
     * type.
     *
     * @param name The name of the command interceptor.
     * @param event The event being intercepted.
     */
    private static void interceptWith(String name, NexusCommandEvent event) {
        try {
            NexusCommandInterceptor interceptor = interceptors.getOrDefault(name, null);

            if (interceptor == null) {
                return;
            }

            if (interceptor instanceof NexusMiddleware) {
                ((NexusMiddleware) interceptor).onBeforeCommand(new NexusMiddlewareEventCore(event));
            } else if (interceptor instanceof NexusAfterware){
                ((NexusAfterware) interceptor).onAfterCommandExecution(event);
            }
        } catch (Exception exception) {
            exception.printStackTrace();
        }
    }

    /**
     * An internal method that assesses multiple command interceptors and expects a
     * proper boolean response.
     *
     * @param names The names of the command interceptors to execute.
     * @param event The event to execute.
     * @return Are all interceptors agreeing with the command execution?
     */
    public static NexusMiddlewareGate interceptWithMany(List<String> names, NexusCommandEvent event) {
        // This is intentionally a for-loop since we want to stop at a specific point.
        NexusMiddlewareGateCore gate = NexusMiddlewareGateRepository.get(event.getBaseEvent().getInteraction());
        for (String name : names) {
            interceptWith(name, event);
            if (!gate.allowed()) {
                return NexusMiddlewareGateRepository.release(event.getBaseEvent().getInteraction());
            }
        }

        NexusMiddlewareGateRepository.release(event.getBaseEvent().getInteraction());
        return null;
    }

}

File: src/main/java/pw/mihou/nexus/features/command/interceptors/core/NexusCommandInterceptorCore.kt
package pw.mihou.nexus.features.command.interceptors.core

import pw.mihou.nexus.features.command.interceptors.facades.NexusAfterware
import pw.mihou.nexus.features.command.interceptors.facades.NexusCommandInterceptor
import pw.mihou.nexus.features.command.interceptors.facades.NexusMiddleware
import pw.mihou.nexus.Nexus
import pw.mihou.nexus.features.command.core.NexusMiddlewareEventCore
import pw.mihou.nexus.features.command.facade.NexusCommandEvent
import pw.mihou.nexus.features.command.validation.middleware.OptionValidationMiddleware

internal object NexusCommandInterceptorCore {

    private val interceptors: MutableMap<String, NexusCommandInterceptor> = mutableMapOf(
        OptionValidationMiddleware.NAME to OptionValidationMiddleware
    )

    /**
     * Adds one middleware to [Nexus].
     * @param name the name of the middleware.
     * @param middleware the middleware to add.
     */
    @JvmStatic
    fun addMiddleware(name: String, middleware: NexusMiddleware) {
        interceptors[name] = middleware
    }

    /**
     * Adds one afterware to [Nexus].
     * @param name the name of the afterware.
     * @param afterware the afterware to add.
     */
    @JvmStatic
    fun addAfterware(name: String, afterware: NexusAfterware) {
        interceptors[name] = afterware
    }

    @JvmStatic
    fun has(name: String) = interceptors.containsKey(name)

    @JvmStatic
    fun hasAfterware(name: String) = interceptors.containsKey(name) && interceptors[name] is NexusAfterware


    @JvmStatic
    fun hasMiddleware(name: String) = interceptors.containsKey(name) && interceptors[name] is NexusMiddleware

    @JvmStatic
    fun middlewares(names: List<String>): Map<String, NexusMiddleware> = names
        .map { it to interceptors[it] }
        .filter { it.second != null && it.second is NexusMiddleware }
        .associate { it.first to it.second as NexusMiddleware }

    internal fun afterwares(names: List<String>): List<NexusAfterware> = names
        .map { interceptors[it] }
        .filter { it != null && it is NexusAfterware }
        .map { it as NexusAfterware }

    @JvmStatic
    fun execute(event: NexusCommandEvent, middlewares: Map<String, NexusMiddleware>): NexusMiddlewareGateCore? {
        val gate = NexusMiddlewareGateCore()
        for ((name, middleware) in middlewares) {
            try {
                middleware.onBeforeCommand(NexusMiddlewareEventCore(event, gate))
                if (!gate.isAllowed) {
                    event.store(NexusAfterware.BLOCKING_MIDDLEWARE_KEY, name)
                    return gate
                }
            } catch (exception: Exception) {
                Nexus.logger.error("An uncaught exception was caught while trying to execute a middleware.", exception)
            }
        }
        return null
    }

    internal fun execute(event: NexusCommandEvent, afterwares: List<NexusAfterware>, dispatched: Boolean = true) {
        for (afterware in afterwares) {
            try {
                if (dispatched) {
                    afterware.onAfterCommandExecution(event)
                } else {
                    afterware.onFailedDispatch(event)
                }
            } catch (exception: Exception) {
                Nexus.logger.error("An uncaught exception was caught while trying to execute an afterware.", exception)
            }
        }
    }

}